src/main/java/org/datadog/jenkins/plugins/datadog/listeners/DatadogGraphListener.java

```diff
@@ -0,0 +1,158 @@
+import org.jenkinsci.plugins.workflow.actions.StageAction;
+import org.jenkinsci.plugins.workflow.actions.ThreadNameAction;
+import org.jenkinsci.plugins.workflow.actions.TimingAction;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepEndNode;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.flow.GraphListener;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+
+/**
+ * A GraphListener implementation which computes timing information
+ * for the various stages in a pipeline.
+ */
+@Extension
+public class DatadogGraphListener implements GraphListener {
+
+    private static final Logger logger = Logger.getLogger(DatadogGraphListener.class.getName());
+
+    @Override
+    public void onNewHead(FlowNode flowNode) {
+        if (!isMonitored(flowNode)) {
+            return;
+        }
+        StepEndNode endNode = (StepEndNode) flowNode;
+        StepStartNode startNode = endNode.getStartNode();
+        int stage_depth = 0;
+        for (BlockStartNode node : startNode.iterateEnclosingBlocks()) {
+            if (isStageNode(node)) {
+                stage_depth++;
+            }
+        }
+        DatadogClient client = ClientFactory.getClient();
+        try {
+            BuildData buildData = new BuildData(getRun(flowNode), flowNode.getExecution().getOwner().getListener());
+            String hostname = buildData.getHostname("");
+            Map<String, Set<String>> tags = buildData.getTags();
+            TagsUtil.addTagToTags(tags, "stage_name", getStageName(startNode));
+            TagsUtil.addTagToTags(tags, "stage_depth", String.valueOf(stage_depth));
+            tags.remove("result"); // Jenkins sometimes consider the build has a result even though it's still running.
+            client.gauge("jenkins.job.stageduration", getTime(startNode, endNode), hostname, tags);
+        } catch (IOException | InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private boolean isMonitored(FlowNode flowNode) {
+        // Filter the node out if it is not the end of step
+        // Timing information is only available once the step has completed.
+        if (!(flowNode instanceof StepEndNode)) {
+            return false;
+        }
+
+        // Filter the node if the job has been blacklisted from the Datadog plugin configuration.
+        WorkflowRun run = getRun(flowNode);
+        if (run == null || !DatadogUtilities.isJobTracked(run.getParent().getFullName())) {
+            return false;
+        }
+
+        // Filter the node out if it is not the end of a stage.
+        // The plugin only monitors timing information of stages
+        return isStageNode(((StepEndNode) flowNode).getStartNode());
+
+        // Finally return true as this node is the end of a monitored stage.
```

Comment:
These comments are a little confusing. Should we first check `isStageNode` in an if statement and then `return true` by default at the end?

Reply:
Oh yeah, my IDE removing if blocks for optimization. Will fix